bin/monitor

```diff
@@ -37,6 +37,9 @@ loop do
         pulse = r.check_pulse(session: sessions[r.id], previous_pulse: pulse) if sessions[r.id]
         Clog.emit("Got new pulse") { {got_pulse: {ubid: r.ubid, pulse: pulse}} }
         sleep r.monitoring_interval
+      rescue => ex
```

Comment:
You sure this is the exception type you want to grab? I always have to consult the manual, but you probably want "NoMethodError" to actually crash the program (deleterious though it may be, it tends to spin), whereas indeed, most kinds of IOError and some others others should be logged only.

Reply:
Looking at the list [here](https://ruby-doc.org/core-2.5.1/Exception.html), I think we would want to retry anything under StandardError (which is what we are grabbing here) except ArgumentError, NameError, NoMethodError and FrozenError. I updated the code, what do you think about this list? Should we add/remove anything?

Reply:
I think I'd go the other way and rescue:

RuntimeError
IOError
and *maybe* SystemCallError. But perhaps not even that to start.

https://docs.ruby-lang.org/en/3.2/Exception.html

I think there's a few exceptions you'd be spinning on you are missing with your itemization; really the cases we positively want to retry are pretty narrow in my experience.

Though, unlike respirate's heartbeat, we don't have meta-monitoring on pulses. I suggest adding some to catch any unusual crashes rather than trying to make the exception handling internally airtight (there are ways to die without an exception). A simple way is to check that the database connection is live, but there are ways to do even better.

Reply:
For now, I catch all StandardError, retry RuntimeError and IOError and log the other ones; we can use that log to potentially decide to retry more.